stride/utils/operators.py

```diff
@@ -29,27 +29,117 @@ def adjoint(self, d_mul, a, b, **kwargs):
         return d_a, d_b
 
 
+@mosaic.tessera
 class Concatenate(Operator):
-
-    def forward(self, *args, **kwargs):
-        axis = kwargs.pop('axis', 0)
+    """
+    Concatenate multiple StructuredData objects.
+
+    Parameters
+    ----------
+    start_end: 2d-array, optional
+        Array containing start and end indices pointing to each objects' position in the concatenated object.
+        Shape should be (num_objects, 2)
+    new_axis : bool, optional
+        Whether to concatenate on a new axis. Defaults to False.
+    axis : int, optional
+        Axis to perform the concatenation. Defaults to 0.
+
+    """
+    def __init__(self, *args, **kwargs):
+        self._start_end = kwargs.pop('start_end', None)
+        self.new_axis = kwargs.pop('new_axis', False)
+        self.axis = kwargs.pop('axis', 0)
+        super().__init__(*args, **kwargs)
+
+    async def forward(self, *args, **kwargs):
+        """
+        Parameters
+        ----------
+        args : Sequence[StructuredData]
+            Sequence of StructuredData objects to be combined.
+        start_end : 2d-array, optional
+            Array containing start and end indices pointing to each objects' position in the concatenated object.
+            Shape should be (num_objects, 2)
+        axis : int, optional
+            The axis to concatenate on. Defaults to 0.
+        new_axis : bool, optional
+            Whether to create a new axis when concatenating, or maintain the dimensions. Defaults to False.
+
+        Returns
+        -------
+        StructuredData
+            Concatenated data as a single StructuredData object.
+
+        """
+        new_axis = kwargs.pop('new_axis', None)
+        axis = kwargs.pop('axis', None)
+        start_end = kwargs.pop('start_end', None)
+
+        if new_axis is not None:
+            self.new_axis = new_axis
+
+        if axis is not None:
+            self.axis = axis
+
+        if start_end is not None:
+            # update if required
+            self._start_end = start_end
+        elif self._start_end is None:
+            # build start_end from the arguments
+            self.build_start_end(args)
 
         concat_data = [np.array(each.data) for each in args]
```

Comment:
is the `np.array(...)` needed? `each.data` should already be an array, right?

Reply:
Completed